src/kudu/consensus/replicate_msg_wrapper.h

```diff
@@ -0,0 +1,190 @@
+#include "kudu/consensus/ref_counted_replicate.h"
+#include "kudu/gutil/strings/substitute.h"
+#include "kudu/util/compression/compression.pb.h"
+#include "kudu/util/compression/compression_codec.h"
+#include "kudu/util/crc.h"
+#include "kudu/util/faststring.h"
+
+namespace kudu {
+namespace consensus {
+
+/**
+ * Thin wrapper to handle compression/decompression of replicate msg
+ *
+ * Pass any msg (compressed or uncompressed) to the constructor and then call
+ * Init() to populate both the compressed and uncompressed msgs.
+ */
+class ReplicateMsgWrapper {
+public:
+  ReplicateMsgWrapper(const ReplicateRefPtr &msg,
+                      const CompressionCodec *codec_hint) {
+    orig_msg_ = msg;
+    const CompressionType msg_codec_type =
+        orig_msg_->get()->write_payload().compression_codec();
+    if (msg_codec_type == NO_COMPRESSION) {
+      msg_ = orig_msg_;
+      codec_ = codec_hint;
+    } else {
+      compressed_msg_ = orig_msg_;
+      CHECK_OK(GetCompressionCodec(msg_codec_type, &codec_));
+    }
+    should_compress_ =
```

Comment:
should_compress_ can get set to true, for even messages that are already compressed, and if CompressMsg gets called, it can inadvertently double compress the msg.